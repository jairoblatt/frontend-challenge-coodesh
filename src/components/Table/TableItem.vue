<template>
  <tr class="table-item">
    <td class="item__avatar">
      <Avatar v-if="avatar" size="2.1" border="1.5" class="mr-2">
        <img :src="avatar" :alt="`${nameResolve} avatar`" />
      </Avatar>
      {{ nameResolve }}
    </td>
    <td>{{ gender }}</td>
    <td>{{ age }}</td>
    <td><slot /></td>
  </tr>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import { Nullable, UserName } from "@/types/types";

export default Vue.extend({
  props: {
    gender: String as PropType<string>,
    age: Number as PropType<number>,
    name: {
      type: Object,
      default: null,
    } as PropOptions<Nullable<UserName>>,
    avatar: String as PropType<string>,
  },

  computed: {
    nameResolve(): string {
      const { name } = this;
      return name ? `${name.first} ${name.last}` : "Unknow name";
    },
  },
});
</script>

<style lang="postcss" scoped>
.table-item td {
  @apply py-3 px-2 text-light-onSurfacePrimary;
}

.item__avatar {
  @apply flex items-center;
}
</style>
