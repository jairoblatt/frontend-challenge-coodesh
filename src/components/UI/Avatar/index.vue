<template>
  <component :is="tag" class="avatar" :style="styles">
    <slot />
  </component>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Avatar",

  props: {
    tag: {
      type: String,
      default: "div",
    } as PropOptions<string>,
    // The prop size it's definedly with 'rem' css unit meansure
    size: {
      type: [String, Number],
      default: 2,
    } as PropOptions<string | number>,
    // The prop border it's definedly with 'px' css unit meansure
    border: [String, Number] as PropType<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        height: `${this.size}rem`,
        width: `${this.size}rem`,
        borderWidth: `${this.border || Number(this.size) / 2}px`,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.avatar {
  @apply rounded-full border-4 border-primary-base border-solid overflow-hidden;
}

.avatar img,
source {
  @apply min-w-full h-auto object-contain;
}

/* absolute -top-28 left-2/4 transform -translate-x-2/4 */

/* .user__avatar img,
source {
  height: 12rem;
  width: 12rem;
} */
</style>
