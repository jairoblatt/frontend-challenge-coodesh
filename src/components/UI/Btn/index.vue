<template>
  <button v-on="$listeners" :class="classes">
    <spinner v-show="loading" size="21" />
    <slot v-show="!loading" />
  </button>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";

export default Vue.extend({
  name: "Btn",

  props: {
    primary: Boolean as PropType<boolean>,
    secondary: Boolean as PropType<boolean>,
    block: Boolean as PropType<boolean>,
    loading: Boolean as PropType<boolean>,
    small: Boolean as PropType<boolean>,
    outlined: Boolean as PropType<boolean>,
  },

  computed: {
    classes() {
      return {
        btn: true,
        "btn--primary": this.primary,
        "btn--secondary": this.secondary,
        "btn--block": this.block,
        "btn--small": this.small,
        "btn--outlined": this.outlined,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.btn {
  @apply rounded-md px-5 py-3 text-sm outline-none font-semibold text-white transition-colors;
}

.btn--primary {
  @apply bg-primary-base hover:bg-primary-light;
}

.btn--secondary {
  @apply bg-secondary-gray;
}

.btn--block {
  @apply w-full;
}

.btn:disabled {
  @apply bg-gray-300 text-gray-900 cursor-auto;
}

.btn--small {
  @apply px-7 py-2;
}

.btn--outlined {
  @apply border border-gray-100 text-light-onSurfacePrimary hover:border-black hover:text-black;
}
</style>
