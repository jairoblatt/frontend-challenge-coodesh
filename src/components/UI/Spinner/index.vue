<template>
  <div class="spinner-container">
    <span class="spinner" :style="styles"></span>
  </div>
</template>
<script lang="ts">
import Vue, { PropOptions } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Spinner",

  props: {
    size: {
      type: [String, Number],
      default: "inherit",
    } as PropOptions<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        minHeight: `${this.size}px`,
        minWidth: `${this.size}px`,
      };
    },
  },
});
</script>
<style lang="postcss" scoped>
.spinner-container {
  @apply relative flex justify-center items-center min-h-full min-w-full;
}

.spinner {
  @apply pointer-events-none border-solid rounded-full  border-primary-base  p-1;

  border-width: 0.3rem;
  border-top-color: white;
  animation: spinner 0.5s linear infinite;
}

@keyframes spinner {
  100% {
    transform: rotate(360deg);
  }
}
</style>
