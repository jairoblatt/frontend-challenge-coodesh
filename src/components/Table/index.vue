<template>
  <table class="table__wrapper">
    <thead>
      <tr>
        <th>Name</th>
        <th>Gender</th>
        <th>Birth</th>
        <th>Actions</th>
      </tr>
    </thead>

    <transition-group tag="tbody" name="slide-down">
      <TableItem
        v-for="(user, index) in users"
        :key="`table-item-${index}`"
        :name="user.name"
        :gender="user.gender"
        :age="user.dob.age"
        :avatar="user.picture.thumbnail"
      >
        <Btn outlined small @click="redirectToUserPage(user)">
          View
        </Btn>
      </TableItem>
    </transition-group>
  </table>
</template>

<script lang="ts">
import Vue, { PropOptions } from "vue";
import TableItem from "./TableItem.vue";
import { UserData } from "@/types/IUser";

export default Vue.extend({
  components: {
    TableItem,
  },

  props: {
    users: {
      type: Array,
      default: () => [],
    } as PropOptions<UserData[]>,
  },

  methods: {
    redirectToUserPage(user: UserData) {
      this.$router.push({
        name: "user",
        params: { username: user.login.username },
        query: this.$route.query,
      });
    },
  },
});
</script>

<style lang="postcss" scoped>
.table__wrapper {
  @apply rounded-sm border border-gray-100 overflow-hidden text-center bg-white border-collapse w-full;
}

.table__wrapper thead tr {
  @apply text-light-onSurfacePrimary;
}

.table__wrapper th {
  @apply uppercase font-bold;

  padding: 1rem 2rem;
  letter-spacing: 0.1rem;
  font-size: 0.7rem;
}

.table__wrapper tr:nth-child(even) {
  @apply bg-gray-50 bg-opacity-60;
}

.table__wrapper tbody tr {
  @apply border border-gray-100 border-solid;
}
</style>
