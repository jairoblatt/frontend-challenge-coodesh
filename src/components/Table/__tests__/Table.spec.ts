import { mount, Wrapper, MountOptions, createLocalVue } from "@vue/test-utils";
import Table from "../index.vue";
import TableItem from "../TableItem.vue";
import Btn from "@/components/UI/Btn/index.vue";
import Spinner from "@/components/UI/Spinner/index.vue";
import Avatar from "@/components/UI/Avatar/index.vue";

const localVue = createLocalVue();
localVue.component("TableItem", TableItem);
localVue.component("Btn", Btn);
localVue.component("Spinner", Spinner);
localVue.component("Avatar", Avatar);

type Instance = InstanceType<typeof Table>;
let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

const $router = {
  push: jest.fn(),
};

const $route = {
  query: "?seed=123asd&page=2",
};

function fakeUsers() {
  return Array.from({ length: 5 }, () => ({
    name: {
      first: "Name",
      last: "Here",
    },
    gender: "Male",
    dob: {
      age: 20,
    },
    login: {
      username: "my-login",
    },
    picture: {
      thumbnail: "avatar.png",
    },
  }));
}

describe("Table.vue ", () => {
  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(Table, {
        ...options,
        localVue,
        stubs: {
          Transition: false,
        },
        mocks: {
          $router,
          $route,
        },
      });
    };
  });

  it("Should mount the component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should render the tableItem component when has user", async () => {
    const wrapper = mountFunction({
      propsData: {
        users: fakeUsers(),
      },
    });

    expect(wrapper.findComponent(TableItem).exists()).toBe(true);
  });

  it("Should render 5 tabItem based in users prop", () => {
    const wrapper = mountFunction({
      propsData: {
        users: fakeUsers(),
      },
    });

    expect(wrapper.findAllComponents(TableItem)).toHaveLength(5);
  });

  it("Should redirect route when a user is clicked", () => {
    const users = fakeUsers();

    const wrapper = mountFunction({
      propsData: {
        users,
      },
    });

    const btnComponent = wrapper.findAllComponents(Btn).at(0);
    btnComponent.trigger("click");

    expect($router.push).toHaveBeenCalled();
    expect($router.push).toHaveBeenCalledWith({
      name: "user",
      params: {
        username: users[0].login.username,
      },
      query: $route.query,
    });
  });
});
