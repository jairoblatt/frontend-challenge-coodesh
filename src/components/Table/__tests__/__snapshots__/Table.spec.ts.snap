// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.vue  Should mount the component 1`] = `
"<table class=\\"table__wrapper\\">
  <thead>
    <tr>
      <th>Name</th>
      <th>Gender</th>
      <th>Birth</th>
      <th>Actions</th>
    </tr>
  </thead>
  <transition-group-stub tag=\\"tbody\\" name=\\"slide-down\\"></transition-group-stub>
</table>"
`;
